enterprise/server/util/cpuset/cpuset.go

```diff
@@ -0,0 +1,91 @@
+package cpuset
+
+import (
+	"sync"
+
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/resources"
+	"github.com/buildbuddy-io/buildbuddy/server/util/priority_queue"
+	"golang.org/x/exp/slices"
+)
+
+const emptyValue = -1
+
+// Compile-time check that cpuLeaser implements the interface.
+var _ interfaces.CPULeaser = (*cpuLeaser)(nil)
+
+type cpuLeaser struct {
+	mu     sync.Mutex
+	leases map[int][]string
+}
+
+// Acquire leases a set of CPUs (identified by index) for a task. The returned
+// function should be called to free the CPUs when they are no longer used.
+func (l *cpuLeaser) Acquire(numCPUs int, taskID string) ([]int, func()) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	pq := priority_queue.New[int](priority_queue.WithEmptyValue(emptyValue))
+
+	for cpuid, tasks := range l.leases {
+		// we want the least loaded cpus first, so give the
+		// cpus with more tasks a more negative score.
+		pq.Push(cpuid, -1*len(tasks))
+	}
+
+	leastLoaded := make([]int, 0)
+	for i := 0; i < numCPUs; i++ {
+		cpuid := pq.Pop()
```

Comment:
ack, the downside is that it makes the API for Peek() and Pop() noisier, and because it's not a built in type like map we don't get to have it both ways.